File: python/zebra-puzzle/zebra_puzzle.py
from collections import namedtuple
from enum import IntEnum, StrEnum
from itertools import permutations


def drinks_water():
    return _find_house(solution(), Drink.WATER).nationality


def owns_zebra():
    return _find_house(solution(), Pet.ZEBRA).nationality


class Solution:
    def __init__(self):
        self.cached = None

    def get(self):
        if self.cached is None:
            self.cached = generate()
        return self.cached


solution = Solution().get


def generate():
    return next(
        _make_houses(colors, nats, pets, drinks, hobbies)
        for nats in permutations(Nationality)
        if _check_rule_10(nats)
        for colors in permutations(Color)
        if _check_rule_6(colors) and _check_rule_2(colors, nats) and _check_rule_15(colors, nats)
        for drinks in permutations(Drink)
        if _check_rule_9(drinks) and _check_rule_5(nats, drinks) and _check_rule_4(colors, drinks)
        for hobbies in permutations(Hobby)
        if _check_rule_8(colors, hobbies) and _check_rule_13(drinks, hobbies) and _check_rule_14(nats, hobbies)
        for pets in permutations(Pet)
        if _check_rule_3(nats, pets) and _check_rule_7(pets, hobbies) and _check_rule_11(pets, hobbies) and _check_rule_12(pets, hobbies)
    )


House = namedtuple("House", "color, nationality, pet, drink, hobby")


def _make_houses(colors, nats, pets, drinks, hobbies):
    return [House(*attrs) for attrs in zip(colors, nats, pets, drinks, hobbies)]


def _find_house(houses, attr):
    return next(
        house for house in houses
        if getattr(house, attr.__class__.__name__.lower()) == attr.value)


class Color(IntEnum):
    RED = 0
    GREEN = 1
    IVORY = 2
    YELLOW = 3
    BLUE = 4


class Nationality(StrEnum):
    ENG = "Englishman"
    SPA = "Spaniard"
    UKR = "Ukranian"
    NOR = "Norwegian"
    JAP = "Japanese"


class Pet(IntEnum):
    DOG = 0
    SNAIL = 1
    FOX = 2
    HORSE = 3
    ZEBRA = 4


class Drink(IntEnum):
    COFFEE = 0
    TEA = 1
    MILK = 2
    JUICE = 3
    WATER = 4


class Hobby(IntEnum):
    DANCING = 0
    PAINTER = 1
    READING = 2
    CHESS = 3
    FOOTBALL = 4


# Rule 1: There are five houses.
# We construct the solution this way.


def _check_rule_2(colors, nats):
    """The Englishman lives in the red house."""
    return all((nats[i] == Nationality.ENG) == (colors[i] == Color.RED)
               for i in range(5))


def _check_rule_3(nats, pets):
    """The Spaniard owns the dog."""
    return all((nats[i] == Nationality.SPA) == (pets[i] == Pet.DOG)
               for i in range(5))


def _check_rule_4(colors, drinks):
    """The person in the green house drinks coffee."""
    return all((colors[i] == Color.GREEN) == (drinks[i] == Drink.COFFEE)
               for i in range(5))


def _check_rule_5(nats, drinks):
    """The Ukrainian drinks tea."""
    return all((nats[i] == Nationality.UKR) == (drinks[i] == Drink.TEA)
               for i in range(5))


def _check_rule_6(colors):
    """The green house is immediately to the right of the ivory house."""
    return (i:=colors.index(Color.GREEN)) > 0 and colors[i-1] == Color.IVORY


def _check_rule_7(pets, hobbies):
    """The snail owner likes to go dancing."""
    return all((pets[i] == Pet.SNAIL) == (hobbies[i] == Hobby.DANCING)
               for i in range(5))


def _check_rule_8(colors, hobbies):
    """The person in the yellow house is a painter."""
    return all((colors[i] == Color.YELLOW) == (hobbies[i] == Hobby.PAINTER)
               for i in range(5))


def _check_rule_9(drinks):
    """The person in the middle house drinks milk."""
    return drinks[2] == Drink.MILK


def _check_rule_10(nats):
    """The Norwegian lives in the first house."""
    return nats[0] == Nationality.NOR


def _check_rule_11(pets, hobbies):
    """The person who enjoys reading lives in the house next to the person with the fox."""
    return abs(hobbies.index(Hobby.READING) - pets.index(Pet.FOX)) == 1


def _check_rule_12(pets, hobbies):
    """The painter's house is next to the house with the horse."""
    return abs(hobbies.index(Hobby.PAINTER) - pets.index(Pet.HORSE)) == 1


def _check_rule_13(drinks, hobbies):
    """The person who plays football drinks orange juice."""
    return drinks[hobbies.index(Hobby.FOOTBALL)] == Drink.JUICE


def _check_rule_14(nats, hobbies):
    """The Japanese person plays chess."""
    return hobbies[nats.index(Nationality.JAP)] == Hobby.CHESS


def _check_rule_15(colors, nats):
    """The Norwegian lives next to the blue house."""
    return abs(nats.index(Nationality.NOR) - colors.index(Color.BLUE)) == 1

File: haskell/zebra-puzzle/src/ZebraPuzzle.hs
module ZebraPuzzle (Resident (..), Solution (..), solve) where

import Data.List (permutations, zipWith5)

data Resident
  = Englishman
  | Spaniard
  | Ukrainian
  | Norwegian
  | Japanese
  deriving (Eq, Show, Enum, Bounded)

data Solution = Solution
  { waterDrinker :: Resident,
    zebraOwner :: Resident
  }
  deriving (Eq, Show)

solve :: Solution
solve =
  Solution
    { waterDrinker = residentWho ((Water ==) . drink) solution,
      zebraOwner = residentWho ((Zebra ==) . pet) solution
    }
  where
{- ORMOLU_DISABLE -}
    solution = head $ [
      zipWith5 House rs cs ps ds hs | rs <- permutations (enumerate::[Resident])
                                    , rule10 rs
                                    , cs <- permutations (enumerate::[Color])
                                    , rule6  cs
                                    , rule2  cs rs
                                    , rule15 cs rs
                                    , ds <- permutations (enumerate::[Drink])
                                    , rule9  ds
                                    , rule5  rs ds
                                    , rule4  cs ds
                                    , hs <- permutations (enumerate::[Hobby])
                                    , rule8  cs hs
                                    , rule13 ds hs
                                    , rule14 rs hs
                                    , ps <- permutations (enumerate::[Pet])
                                    , rule3  rs ps
                                    , rule7  ps hs
                                    , rule11 ps hs
                                    , rule12 ps hs
                                    ]
{- ORMOLU_ENABLE -}

data House = House
  { resident :: Resident,
    color :: Color,
    pet :: Pet,
    drink :: Drink,
    hobby :: Hobby
  }
  deriving (Eq, Show)

data Color = Red | Green | Ivory | Yellow | Blue deriving (Eq, Show, Enum, Bounded)

data Pet = Dog | Snail | Fox | Horse | Zebra deriving (Eq, Show, Enum, Bounded)

data Drink = Coffee | Tea | Milk | Juice | Water deriving (Eq, Show, Enum, Bounded)

data Hobby = Dancing | Paint | Reading | Chess | Football deriving (Eq, Show, Enum, Bounded)

-- Rule 1: # Rule 1: There are five houses.

-- | The Englishman lives in the red house.
rule2 :: [Color] -> [Resident] -> Bool
rule2 = checkPaired Red Englishman

-- | The Spaniard owns the dog.
rule3 :: [Resident] -> [Pet] -> Bool
rule3 = checkPaired Spaniard Dog

-- | The person in the green house drinks coffee.
rule4 :: [Color] -> [Drink] -> Bool
rule4 = checkPaired Green Coffee

-- | The Ukrainian drinks tea.
rule5 :: [Resident] -> [Drink] -> Bool
rule5 = checkPaired Ukrainian Tea

-- | The green house is immediately to the right of the ivory house.
rule6 :: [Color] -> Bool
-- Red is some initial value on the right that isn't Green. I do this differently in other rules.
rule6 = fst . foldr (\cur (rez, prev) -> (rez || prev == Green && cur == Ivory, cur)) (False, Red)

-- | The snail owner likes to go dancing.
rule7 :: [Pet] -> [Hobby] -> Bool
rule7 = checkPaired Snail Dancing

-- | The person in the yellow house is a painter.
rule8 :: [Color] -> [Hobby] -> Bool
rule8 = checkPaired Yellow Paint

-- | The person in the middle house drinks milk.
rule9 :: [Drink] -> Bool
rule9 ds = ds !! 2 == Milk

-- | The Norwegian lives in the first house.
rule10 :: [Resident] -> Bool
rule10 rs = rs !! 0 == Norwegian {- HLINT ignore "Use head" -}

-- | The person who enjoys reading lives in the house next to the person with the fox.
rule11 :: [Pet] -> [Hobby] -> Bool
rule11 = checkNext Fox Reading

-- | The painter's house is next to the house with the horse.
rule12 :: [Pet] -> [Hobby] -> Bool
rule12 = checkNext Horse Paint

-- | The person who plays football drinks orange juice.
rule13 :: [Drink] -> [Hobby] -> Bool
rule13 = checkPaired Juice Football

-- | The Japanese person plays chess.
rule14 :: [Resident] -> [Hobby] -> Bool
rule14 = checkPaired Japanese Chess

-- | The Norwegian lives next to the blue house.
rule15 :: [Color] -> [Resident] -> Bool
rule15 = checkNext Blue Norwegian

checkPaired :: (Eq a, Eq b) => a -> b -> [a] -> [b] -> Bool
checkPaired a b as bs = (a, b) `elem` zip as bs

checkNext :: (Eq a, Eq b) => a -> b -> [a] -> [b] -> Bool
checkNext a b as bs = (a, b) `elem` zip as (tail bs) ++ zip (tail as) bs

enumerate :: (Enum a, Bounded a) => [a]
enumerate = [minBound .. maxBound]

residentWho :: (House -> Bool) -> [House] -> Resident
residentWho f = resident . head . filter f
